### ISSUE TITLE
Possible fixers that we have on fabbot.io but should be part of PHP CS Fixer

### ISSUE THREAD

**Issue by fabpot** [FriendsOfPHP/PHP-CS-Fixer#1208]
I've just removed a couple of fixers I have on fabbot.io because there are now part of the PHP CS fixer (and they are more robust that my simple and stupid regexes).

I have two more of them that should really be part of PHP CS Fixer:

 * fix of inheritdoc tags:

```
'#@inheritdoc#i' => '{@inheritdoc}',
'#{{@inheritdoc}}#i' => '{@inheritdoc}', '#{@inheritDoc}#i' => '{@inheritdoc}',
'#@{inheritdoc}#i' => '{@inheritdoc}',
```

 * fix of PHPUnit constructs (not sure if you want to go down this road though):

```
'{\$this\->assertNot(Equals|Same)\s*\(\s*null\b\s*,\s*}si' => '$this->assertNotNull(',
'{\$this\->assert(Equals|Same)\s*\(\s*null\b\s*,\s*}si' => '$this->assertNull(',
'{\$this\->assert(Equals|Same)\s*\(\s*true\b\s*,\s*}si' => '$this->assertTrue(',
'{\$this\->assert(Equals|Same)\s*\(\s*false\b\s*,\s*}si' => '$this->assertFalse(',
```

**Comment by keradus** [comment]
Great ones! It would be great to have them here

**Comment by GrahamCampbell** [comment]
I can do the inherit phpdoc stuff if you want, but it'll have to be in early July, just like array fixer I was going to do, since that's when I have a nice bit of free time. :)

**Comment by GrahamCampbell** [comment]
> fix of PHPUnit constructs

That looks pretty useful tbh. :+1:

**Comment by keradus** [comment]
BTW
> '{\$this\->assert(Equals|Same)\s*\(\s*true\b\s*,\s*}si' => '$this->assertTrue(',

Even if this will be indeed invokation of pure PHPUnit method then `assertEquals(true)` is way different that `assertSame(true)`. So changing both of them to `assertTrue()` is as dungerous as `strict` fixer. Then whole antipathy that was about `strict` fixer should be about that new fixer as well...

**Comment by keradus** [comment]
I think we shoud distinguish fixer that change `assertEquals` into `assertSame` and fixer that change `assertSame(true, ...)` into `assertTrue(...)`

**Comment by keradus** [comment]
Requested fixers were added in just released v1.10 as:

> * `phpdoc_inline_tag [symfony]`
> Fix PHPDoc inline tags, make inheritdoc always inline.
> 
> * `php_unit_strict [contrib]`
> PHPUnit methods like "assertSame" should be used instead of "assertEquals".
> Warning! This could change code behavior.
> 
> * `php_unit_construct [contrib]`
> PHPUnit assertion method calls like "->assertSame(true, $foo)" should be written with dedicated method like "->assertTrue($foo)".
> Warning! This could change code behavior.
> 

cc @fabpot

**Comment by fabpot** [comment]
Just for the record, the PHPunit fixers added in 1.10 are quite different from the ones requested in this issue. So, I'm going to keep the one I have in fabbot.

Let me explain:

Replacing `assertEquals` by `assertSame` is indeed quite dangerous and can change the behavior. But when someone writes `assertEquals(true, ...)`, this is almost always a bug as he probably really meant `assertSame(true, ...)`, which in turn can be transformed safely to `assertTrue(...)`. So, the decoupling of the two makes both of them dangerous and not interesting at all if you ask me.

To sum up, my proposal was to fix potential bugs in PHPUnit tests where people mistakenly used `assertEquals` when the should have used `assertSame` instead and at the same time to simplify the assertions with more specific PHPUnit methods.

**Comment by keradus** [comment]
And in that case, after using both fixers (in order of priorities) - `assertEquals(true, ...` will be changed into `assertTrue(...`

> '{\$this\->assert(Equals|Same)\s*\(\s*true\b\s*,\s*}si' => '$this->assertTrue(',

If one want to be a little safier - he turns on one fixer and change only `assertSame`, if he one both cases to be changed - he turns on both fixers.

If I get you right @fabpot, you want fixer `php_unit_construct` to fix both `assertSame` and `assertEquals` ? That would be extra easy to change.

**Comment by stof** [comment]
@keradus but we have lots of valid use cases for ``assertEquals``. We don't want to change them all

**Comment by keradus** [comment]
That's why I asked about extending `php_unit_construct` to change
* `assertSame(true, ...)` into `assertTrue(...)`
* `assertEquals(true, ...)` into `assertTrue(...)`
but not
* `assertEquals($a, $b) into `assertSame($a, $b)`

**Comment by fabpot** [comment]
The regexes I've mentioned in my initial description sum up pretty nicely what I think it should do, and probably not more than that. So, your first two examples are indeed good, but not the last one. Put another way, it's only a good idea to change `assertEquals` to `assertSame` when the comparison is with `true`, `false`, or `null`.

**Comment by GrahamCampbell** [comment]
> it's only a good idea to change assertEquals to assertSame when the comparison is with true, false, or null.

Yeh, that's what I was getting at in that this change is really breaking, and makes no sense otherwise.

**Comment by keradus** [comment]
Then please @fabpot @GrahamCampbell take a look at #1332
